pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -189,42 +200,58 @@ func (pl *DefaultPreemption) SelectVictimsOnNode(
 	if status := pl.fh.RunFilterPluginsWithNominatedPods(ctx, state, pod, nodeInfo); !status.IsSuccess() {
 		return nil, 0, status
 	}
-	var victims []*v1.Pod
+
 	numViolatingVictim := 0
-	sort.Slice(potentialVictims, func(i, j int) bool { return util.MoreImportantPod(potentialVictims[i].Pod, potentialVictims[j].Pod) })
+	// Sort potentialVictims by pod priority from high to low, which helps to
+	// reprieve pods with higher priority as many as possible.
+	sort.Slice(potentialVictims, func(i, j int) bool {
+		return util.MoreImportantPod(potentialVictims[i].podInfo.Pod, potentialVictims[j].podInfo.Pod)
+	})
 	// Try to reprieve as many pods as possible. We first try to reprieve the PDB
 	// violating victims and then other non-violating ones. In both cases, we start
 	// from the highest priority victims.
-	violatingVictims, nonViolatingVictims := filterPodsWithPDBViolation(potentialVictims, pdbs)
-	reprievePod := func(pi *framework.PodInfo) (bool, error) {
-		if err := addPod(pi); err != nil {
+	violatingVictims, nonViolatingVictims := filterVictimsWithPDBViolation(potentialVictims, pdbs)
+	reprieveVictim := func(victim *potentialVictim) (bool, error) {
+		if err := addPod(victim.podInfo); err != nil {
 			return false, err
 		}
 		status := pl.fh.RunFilterPluginsWithNominatedPods(ctx, state, pod, nodeInfo)
 		fits := status.IsSuccess()
-		if !fits {
-			if err := removePod(pi); err != nil {
+		if fits {
+			// This pod should be reprieved that higher priority pod can be scheduled without evicting this pod.
+			victim.isReprieved = true
+		} else {
+			// This pod should be preempted.
+			// Remove Pod to simulate this pod has been evicted as a victim.
+			if err := removePod(victim.podInfo); err != nil {
 				return false, err
 			}
-			rpi := pi.Pod
-			victims = append(victims, rpi)
+			rpi := victim.podInfo.Pod
 			logger.V(5).Info("Pod is a potential preemption victim on node", "pod", klog.KObj(rpi), "node", klog.KObj(nodeInfo.Node()))
 		}
 		return fits, nil
 	}
-	for _, p := range violatingVictims {
-		if fits, err := reprievePod(p); err != nil {
+	for _, victim := range violatingVictims {
+		if fits, err := reprieveVictim(victim); err != nil {
 			return nil, 0, framework.AsStatus(err)
 		} else if !fits {
 			numViolatingVictim++
 		}
 	}
 	// Now we try to reprieve non-violating victims.
-	for _, p := range nonViolatingVictims {
-		if _, err := reprievePod(p); err != nil {
+	for _, victim := range nonViolatingVictims {
+		if _, err := reprieveVictim(victim); err != nil {
 			return nil, 0, framework.AsStatus(err)
 		}
 	}
+
+	// Collect all pods that need to be preempted into victims.
+	var victims []*v1.Pod
+	for _, victim := range potentialVictims {
```

Comment:
Since we want to iterate the whole potentialVictims, resorting the victims wouldn't be better, less elements here. Then no need to do the struct refactor anymore.

Reply:
ok, already rollback to sorting way. 

Due to my non-sorting impl is unable to cherry-pick, I rollback to sorting way, and follow the proposal 2 in https://github.com/kubernetes/kubernetes/pull/128307#discussion_r1818265050 to avoid unnecessary sorting.

I am little confuse to figure out for impl of proposal 1, so currently picking proposal 2. Any ideas? @kerthcet @Huang-Wei